### DISCUSSION TITLE
About completions best practice

### DISCUSSION THREAD

**Discussion by tschersach** [zdharma-continuum/zinit#109] [Q&A] [answered]
Hello,

Noobie question, I'm not comfortable with the whole ZSH ecosystem, just a happy user. I'm on a M1 Mac, latest version of `zinit` (new repo)
I have a small question about completions for new tools I install with `brew`.

Below is my full .zshrc.
I have this line `fpath+=( /usr/local/share/zsh/site-functions /opt/homebrew/share/zsh/site-functions )` I can update if I see that tools use a new path for completions. Is updating `fpath` the recommend way to handle completions with zinit? Can I do something better? 

Thanks 😄 

```

### Added by Zinit's installer
if [[ ! -f $HOME/.local/share/zinit/zinit.git/zinit.zsh ]]; then
    print -P "%F{33}▓▒░ %F{220}Installing %F{33}ZDHARMA-CONTINUUM%F{220} Initiative Plugin Manager (%F{33}zdharma-continuum/zinit%F{220})…%f"
    command mkdir -p "$HOME/.local/share/zinit" && command chmod g-rwX "$HOME/.local/share/zinit"
    command git clone https://github.com/zdharma-continuum/zinit "$HOME/.local/share/zinit/zinit.git" && \
        print -P "%F{33}▓▒░ %F{34}Installation successful.%f%b" || \
        print -P "%F{160}▓▒░ The clone has failed.%f%b"
fi

source "$HOME/.local/share/zinit/zinit.git/zinit.zsh"
autoload -Uz _zinit
(( ${+_comps} )) && _comps[zinit]=_zinit

# Load a few important annexes, without Turbo
# (this is currently required for annexes)
zinit light-mode for \
    zdharma-continuum/zinit-annex-rust \
    zdharma-continuum/zinit-annex-as-monitor \
    zdharma-continuum/zinit-annex-patch-dl \
    zdharma-continuum/zinit-annex-bin-gem-node

### End of Zinit's installer chunk

fpath+=( /usr/local/share/zsh/site-functions /opt/homebrew/share/zsh/site-functions )
path+=( $HOME/go/bin $HOME/.cargo/bin )

zinit lucid for \
    atinit"HIST_STAMPS=dd.mm.yyyy" \
    OMZL::history.zsh \
    light-mode pick"async.zsh" src"pure.zsh" \
                sindresorhus/pure

zinit wait lucid for \
    OMZL::completion.zsh \
    atload"
        alias ..='cd ..'
        alias ...='cd ../..'
        alias ....='cd ../../..'
        alias .....='cd ../../../..'
        alias l='ls -lahG'
        alias tf='terraform'    
        export LSCOLORS='Exfxcxdxbxegedabagacad'
        export AWS_DEFAULT_REGION=eu-central-1
	    eval '$(fnm env)'        
    " \
    OMZL::directories.zsh \
	OMZL::git.zsh \
    OMZP::brew \
    OMZP::git \
    OMZL::spectrum.zsh \
	OMZL::termsupport.zsh \
    agkozak/zsh-z \

zinit wait lucid for \
    atinit"zicompinit; zicdreplay" \
    zdharma-continuum/fast-syntax-highlighting \
    atload"_zsh_autosuggest_start" \
    zsh-users/zsh-autosuggestions \
    atload'bindkey "^[[A" history-substring-search-up;
        bindkey "^[[B" history-substring-search-down' \
    zsh-users/zsh-history-substring-search \
    zdharma-continuum/history-search-multi-word \
    blockf atpull'zinit creinstall -q .' \
    zsh-users/zsh-completions
```

**Comment by pschmitt** [comment, marked as answer]
Handling `fpath` is one of the main features of zinit. You can still mess around with it, sure - but I wouldn't advise to do so ;)

To get your custom completions to work I'd just add something like `zinit creinstall -q /opt/homebrew/share/zsh/site-functions` (provided this is the path where brew stores these comps) to either a custom dummy (ie `zdharma-continuum/null`) plugin ~~or directly to the `atinit`of f-sy-h - where you are doing your `zicompinit` (not a good idea since this will run every time you spawn a shell)~~

What you can also do with zinit is have it "manage" brew plugins ie. setup a `null` plugin  which leverages the `atclone`, `atpull` and `atdelete` ices, where you would `brew install $your_apps`, `brew update $your_apps` etc. Then you can make it also call `zi creinstall -q $your_comps_path` in the `atpull` ice for instance, to only update the completions when plugin has received updates.

NB: I never used brew :D

**Reply by paw-lu** [reply]
This was useful! Thanks.

what about migrating functions from explicit `fpath` to zinit?

Before:

```zsh
# Add directory full of functions and autoload them
fpath+=("$HOME/.config/zsh/functions")
autoload -Uz $HOME/.config/zsh/functions/*
```

**Comment by alichtman** [comment]
FWIW I have this in my zshrc. But it sounds like it could use a refactor. https://github.com/alichtman/dotfiles/blob/master/.config/zsh/.zshrc#L152
